blog/en-us/spring-boot-dubbo-start-stop-analysis.md

```diff
@@ -0,0 +1,223 @@
+
+> Note that when we add a breakpoint in Intellij IDEA's Debug mode, we need to use `kill -s INT $PID` or `kill -s TERM $PID` on command line to trigger the breakpoint. The breakpoint will not be trigger if you just click the stop button on the IDE.
+
+You can see the method is called by a thread called `Thread-3`:
+
+```java
+stopAwait:390, StandardServer (org.apache.catalina.core)
+stopInternal:819, StandardServer (org.apache.catalina.core)
+stop:226, LifecycleBase (org.apache.catalina.util)
+stop:377, Tomcat (org.apache.catalina.startup)
+stopTomcat:241, TomcatEmbeddedServletContainer (org.springframework.boot.context.embedded.tomcat)
+stop:295, TomcatEmbeddedServletContainer (org.springframework.boot.context.embedded.tomcat)
+stopAndReleaseEmbeddedServletContainer:306, EmbeddedWebApplicationContext (org.springframework.boot.context.embedded)
+onClose:155, EmbeddedWebApplicationContext (org.springframework.boot.context.embedded)
+doClose:1014, AbstractApplicationContext (org.springframework.context.support)
+run:929, AbstractApplicationContext$2 (org.springframework.context.support)
+```
+
+Through source code analysis, it was executed by Spring's registered `ShutdownHook`.
+
+```java
+    @Override
+    public void registerShutdownHook() {
+        if (this.shutdownHook == null) {
+            // No shutdown hook registered yet.
+            this.shutdownHook = new Thread() {
+                @Override
+                public void run() {
+                    synchronized (startupShutdownMonitor) {
+                        doClose();
+                    }
+                }
+            };
+            Runtime.getRuntime().addShutdownHook(this.shutdownHook);
+        }
+    }
+```
+
+We found that ShutdownHook will execute in the following two cases by referring the Java API documentation[2].
+
+> The Java virtual machine *shuts down* in response to two kinds of events:
+>
+> - The program *exits* normally, when the last non-daemon thread exits or when the `exit` (equivalently, [`System.exit`](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#exit-int-)) method is invoked, or
+> - The virtual machine is *terminated* in response to a user interrupt, such as typing `^C`, or a system-wide event, such as user logoff or system shutdown.
+
+1. call System.exit()
+2. Respond to external signals, such as Ctrl+C(actually sent as SIGINT signal), or `SIGTERM` signal (`kill $PID` will send `SIGTERM` signal by default)
+
+Therefore, the normal application will execute the above ShutdownHook during the stop process (except `kill -9 $PID`). Its function is not only to close the Tomcat, but also to perform other cleanup work, it is unnecessary to go into details.
+
+### Summary
+
+1. Make sure the process can't be exited by start a independency non-daemon thread loop to check the status of variable while `DubboConsumer`startup.
+2. Change the status of variable to make program exit normaly by execute ShutdownHook while `DubboConsumer` exit.
```

Comment:
To stop the `DubboConsumer`, one should call ShutdownHook to change the variable to let the thread break the loop.